### ISSUE TITLE
Sometimes, bubblejail processes remain active/fail to clean up after closing the main window

### ISSUE THREAD

**Issue by odomingao** [igo95862/bubblejail#134]
### Output of `bubblejail --version`

0.9.2

### Your distro name and version

arch

### Description

Sounds similar to #100. The xdg-dbus-proxy process remains active after killing the main window.

Seems pretty random whether the bug happens or not, sometimes it does clean up properly on close.

I need to manually kill the processes, sometimes with SIGKILL because it ignores signals, and then remove the directory in /run/user/1000/bubblejail/

This is usually not a big deal because the waste of resources from the sandbox remaining active is not significant, it only becomes a problem when the sandbox is using an external drive because it prevents things from being unmounted/closed properly on poweroff/reboot.

**Comment by igo95862** [comment]
Hello @odomingao 

Are you sure it is the xdg-dbus-proxy that is holding the bubblejail from terminating?

Can you do a `pstree` whenever the bubblejail fail to terminate?

**Comment by odomingao** [comment]
I apologize for wasting your time, I was mistaken. It was the program itself not closing properly/fully, I probably thought otherwise because I was filtering in htop for the profile name and didn't see the extra processes that were filtered out.

**Comment by odomingao** [comment]
@igo95862 ,

I managed to replicate the issue with a different program. It was actually, in some cases, only xdg-dbus-proxy that remained active 😅. In this case, the problem is that the program failed to launch and this prevents me from launching subsequent instances of it unless I specify the binary to launch. Also note that most of the time it does work normally. It's just sometimes that this happens.

pstree shows:
`  ├─sh───bubblejail───xdg-dbus-proxy───2*[{xdg-dbus-proxy}]`

htop screenshot (searching with / and not filtering..)

![image](https://github.com/user-attachments/assets/7ad8db94-aba5-4ccf-86fc-576df5a2e428)

**Comment by igo95862** [comment]
Interesting.

Would you be able to dump the Python stack with with faulthandler when it happens : https://docs.python.org/3/library/faulthandler.html

There is some clean-up procedures but I guess they are not robust enough.

**Comment by odomingao** [comment]
How can I do this? I still have the process open

**Comment by igo95862** [comment]
I guess if you already have the process running you can only dump its stack which is less informative.

If you can however replicate this hang you can run bubblejail with the `PYTHONFAULTHANDLER=1` environment variable and that will dump the Python stack when it receives the signals like SIGABRT.

**Comment by odomingao** [comment]
Will this save it to a file, or print it to stdout in the terminal in which I kill the program..? Can I get the information even if bubblejail is not launched from a terminal emulator?

**Comment by igo95862** [comment]
It will print it stack to stderr when it receives the signal.

> Can I get the information even if bubblejail is not launched from a terminal emulator?

Depends on how your desktop environment is setup. I am running GNOME which runs each desktop application in its own systemd unit. I can go to the `journalctl` to see what the application printed to stderr.

**Comment by odomingao** [comment]
alright, I'll report back when I have this information

**Comment by igo95862** [comment]
Looks like it is waiting on something? Although not sure why it does not terminate if xdg-dbus-proxy is killed.

**Comment by odomingao** [comment]
Yes, it also ignores signals.

**Comment by igo95862** [comment]
By the way, do you experience the same message when it gets stuck?

```
Traceback (most recent call last):
  File "/usr/lib/bubblejail/bubblejail-helper", line 20, in <module>
    bubblejail_helper_main()
  File "/usr/lib/bubblejail/python-packages/bubblejail/bubblejail_helper.py", line 423, in bubblejail_helper_main
    if "bubblejail-ready" != f.read():
                             ^^^^^^^^
  File "<frozen codecs>", line 321, in decode
TypeError: can't concat NoneType to bytes
```

**Comment by igo95862** [comment]
I think I know how bubblejail gets stuck.

The key is in:

```
File "/usr/lib/bubblejail/python-packages/bubblejail/bubblejail_runner.py", line 291, in read_info_fd
    info_dict = json_load(f)
```

It looks like the the JSON from the info pipe failed to decode. The bubblejail will then be infinitely stuck in the `_run_post_init_hooks` as it will await for the child process id.

I will try to make it more robust.

**Comment by igo95862** [comment]
I am not entirely sure how the `json.decoder.JSONDecodeError: Expecting ',' delimiter: line 2 column 22 (char 23)
` happens but most likely is that bwrap does not write the entire JSON documents atomically and can write fragments of a JSON. I need to add buffering and only load full JSON lines.

**Comment by igo95862** [comment]
@odomingao I pushed commit 42f9227b3dea269156fa666a0bf116a35f7521a9. Can you give it a test? I am not entirely sure it fixed this particular issue but it could have.

**Comment by odomingao** [comment]
Sure, I'll switch to bubblejail-git and report back if the issue still happens

**Comment by igo95862** [comment]
@odomingao have you encountered the issue again after the fix? I plan on releasing the new minor version this weekend if it managed to fix this.

**Comment by odomingao** [comment]
No I haven't

**Comment by igo95862** [comment]
Fixed in 0.9.3
